sys/arm64/arm64/freebsd64_machdep.c

```diff
@@ -332,10 +332,9 @@ freebsd64_sendsig(sig_t catcher, ksiginfo_t *ksi, sigset_t *mask)
 	tf->tf_x[0] = sig;
 	tf->tf_x[1] = (uintcap_t)fp + offsetof(struct sigframe64, sf_si);
 	tf->tf_x[2] = (uintcap_t)fp + offsetof(struct sigframe64, sf_uc);
-
-	trapframe_set_elr(tf, (uintcap_t)catcher);
+	tf->tf_x[8] = (uintcap_t)catcher;
 	tf->tf_sp = (uintcap_t)fp;
-	tf->tf_lr = (register_t)PROC_SIGCODE(p);
+	trapframe_set_elr(tf, cheri_setaddress(tf->tf_elr, PROC_SIGCODE(p)));
```

Comment:
It is not really clear what to do here.  This uses the current user PCC, but if that has restricted bounds this will just fault on return.  We could perhaps initialize p_md.md_sigcode to a capability to use for this during exec and ensure it has the bounds from userspace_root_cap (or derive from userspace_root_cap here).  We have to have large bounds so we can branch to the handler (which can be anywhere in the address space) via an integer branch.  Somewhat related issue before is that the `catcher` cap we use (and that is used as the exception PCC for RISC-V) has to be able to return back to the trampoline so has to have PROC_SIGCODE within its bounds.  Mostly this means that signal handling and any type of narrow PCC bounds in hybrid don't mix well at all without more thought.

Reply:
Making md_sigcode a capability sounds sensible to me. The narrowing PCC bounds in hybrid case isn't new, we already had LR be PROC_SIGCODE and required an integer branch to that to work.

Reply:
(But we should allow the non-signal handler PCC to be arbitrary here and not rely on anything about it)

Reply:
Hm though really you want catcher's bounds etc here, right? Otherwise you will never inherit them.

Reply:
Well, but catcher's bounds might not include the shared page.  Right now we derive catcher's bounds from PCC (tf_elr via __USER_PCC) at the time of freebsd64_sigaction.  The issue is that for hybrid you need the bounds of PCC on return to span both catcher (potentially anywhere in the address space that isn't the shared page) and the sigcode (shared page).  You can't do a full PCC call in the sigcode to try to honor the catcher's bounds since catcher is just going to do a plain integer return that will reuse those bounds to try to return to the sigcode.  This was always true before as well, before we implicitly required the bounds on catcher to include the shared page so that the return from the catcher to the sigcode worked.

Reply:
I guess you would preserve the previous behavior by doing the `cheri_setaddress` on `catcher` instead, but that will just highlight the fact we depend on PCC at the time of sigaction() including bounds for a part of the address space that is generally only "known about" by the kernel.

Reply:
Could we use a capability to the trampoline here and use a capability jump-and-link inside the trampoline to install the `catcher` capability?

Reply:
That won’t work if you turn on PCCBO as the LR sub-register will have the address not the offset

Reply:
I was assuming we update the setting of `catcher` as a real capability holding caller PCC + catcher as the address/offset depending on the PCCBO bit?

Reply:
So this is building a capability to the trampoline (in essence) and `catcher` is already a capability derived from the user PCC at the time of sigaction, but the problem is that the `catcher` routine is not going to do a capability return, but a plain integer return.  And that return is to the second instruction in the sigcode, so even if we did use a capability branch into the `catcher`, the capability we branched into would have to have sufficiently wide bounds to permit the integer return.

Really, this just means that if we ever want to support reduced PCC bounds for hybrid for some kind of compartmentalization, then more work will need to be done to provide sane semantics for signal handlers.  Until such time as we have something concrete to work with though, there's not really much we can do.

I do think I will change this to derive from `catcher` for now though as that will most closely match the existing behavior before this change.